### PR TITLE
server: add --enable-core-dump flag

### LINKED ISSUE #15242: supporting core dumps when Envoy has more capabilities
When Envoy is running inside a container where `CAP_NET_BIND_SERVICE` isn't available but the Envoy binary itself has this capability (e.g.: via `setcap 'cap_net_bind_service=+ep' envoy`), core dumps won't happen unless Envoy was launched via a process with the same capabilities [0].

Forcing deployments to use a helper with bind capabilities to launch Envoy isn't great, because the helper would have to be restricted itself otherwise it could be use by others to gain bind privileges.

One potential solution is to have Envoy re-exec itself, since no new capabilities are gained when Envoy re-executes itself.

Would something like this be acceptable?

cc: @mattklein123 

[0] See PR_SET_DUMPABLE  in https://man7.org/linux/man-pages/man2/prctl.2.html and also https://man7.org/linux/man-pages/man7/capabilities.7.html

### PR DESCRIPTION
**Pull request by rgs1**
This ensures Envoy can core dump when the dumpability bit might have
been unset (e.g.: running inside a container with fewer capabilities
than the ones Envoy itself has).

Fixes #15242.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
I've got `--enable-core-dump` running internally, works great 👌

**Comment by davinci26** [comment]
@rgs1 can you open an issue or add in the existing issue a Windows label so that we know that we have to implement this feature on windows

**Comment by rgs1** [comment]
> @rgs1 can you open an issue or add in the existing issue a Windows label so that we know that we have to implement this feature on windows

What's the exact problem for Windows? Afaik Windows doesn't have capabilities nor a dumpability bit...

**Comment by rgs1** [comment]
> > @rgs1 can you open an issue or add in the existing issue a Windows label so that we know that we have to implement this feature on windows
> 
> What's the exact problem for Windows? Afaik Windows doesn't have capabilities nor a dumpability bit...

Which makes me wonder if we should hide via ifdefs the `--enable-core-dump` command line flag for Windows (and other platforms) without capabilities and this core dumping issue... 

cc: @mattklein123 for thoughts

**Comment by davinci26** [comment]
> What's the exact problem for Windows? Afaik Windows doesn't have capabilities nor a dumpability bit...

We have flag that is not working for one platform so we should probably document it. 

Additionally windows can generate memory dumps from failing applications https://docs.microsoft.com/en-us/windows/client-management/generate-kernel-or-complete-crash-dump

**Comment by rgs1** [comment]
> > What's the exact problem for Windows? Afaik Windows doesn't have capabilities nor a dumpability bit...
> 
> We have flag that is not working for one platform so we should probably document it.

Yeah I agree, I actually wonder if we should hide the flag entirely from platforms for which it isn't implemented.

> Additionally windows can generate memory dumps from failing applications https://docs.microsoft.com/en-us/windows/client-management/generate-kernel-or-complete-crash-dump (the same applies for user mode applications)

Ok, but is this already working out of the box for Windows or is there anything else to be done? In fairness, this `--enable-core-dump` is not the entire truth for Linux given that it still won't be enough if you didn't set `ulimit` correctly and furthermore it isn't really needed if you didn't do anything that won't unset the dumpable bit of the process. Hmm, maybe it should be named `--enable-dumpable-bit` to be more precise. 

> edit: I am not sure if it makes sense to push this functionality to envoy codebase or to the docker image ( see [docs](https://docs.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps)). IMHO the latter makes sense more to me but either one should be doable

You are talking about the Windows case not this PR for the Linux case?

**Comment by davinci26** [comment]
> You are talking about the Windows case not this PR for the Linux case?

Windows case

File: docs/root/operations/cli.rst
Comment (mattklein123) [inline docs/root/operations/cli.rst:358]: Can you add more details about what this actually does and what platforms it works on?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Yeah I agree that if an option only applies to certain platforms we should document it as such and only expose it on those platforms, optimally without a ton of different ifdefs (e.g. using platform capabilities). Thanks for working on this.

/wait

File: docs/root/operations/cli.rst
Comment (rgs1) [reply docs/root/operations/cli.rst]: Done

**Comment by rgs1** [comment]
@mattklein123 this is ready for another pass, I moved the platform specific bits into their own files and reduced the use of ifdefs to a minimum (used in one test). Thanks!

**Comment by rgs1** [comment]
Ah, the bazel bits to split Linux from rest of Posix I am less sure about...

File: docs/root/version_history/current.rst
Comment (mattklein123) [inline docs/root/version_history/current.rst:124]: nit: `:option:` link to option.

File: source/exe/main_common.cc
Comment (mattklein123) [inline source/exe/main_common.cc:57]: If it wasn't actually enabled can we fail and say it's not supported?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks at a high level this LGTM. I will defer to @lizan on the bazel-fu. :)

/wait

File: source/exe/main_common.cc
Comment (rgs1) [reply source/exe/main_common.cc:57]: What would failing mean here? Throwing an exception and exiting? Hmm I am somehow think it should be best effort...

File: docs/root/version_history/current.rst
Comment (rgs1) [reply docs/root/version_history/current.rst:124]: sure adding

File: source/exe/main_common.cc
Comment (mattklein123) [reply source/exe/main_common.cc:57]: Yeah fair enough. A soft failure is fine. Can we at least split the logic and warn if it can't be enabled vs. info for enabled?

/wait

**Comment by rgs1** [comment]
I actually had that but got rid of it to game the coverage failure 😂

I'll add info/warn back.

File: source/exe/main_common.cc
Comment (mattklein123) [inline source/exe/main_common.cc:61]: nit: I don't think you need to log ret now. Also, can you cover both of these cases in tests? It should be pretty easy.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM modulo small nit, thanks!

/wait

File: source/exe/main_common.cc
Comment (rgs1) [reply source/exe/main_common.cc]: Sure np

File: test/server/server_test.cc
Comment (mattklein123) [inline test/server/server_test.cc:1337]: Can you add a no dump enabled test and check for the warning log? Should be very easy.

/wait

Comment (rgs1) [reply test/server/server_test.cc:1337]: yep yep, just pushed that

Comment (rgs1) [reply test/server/server_test.cc:1337]: [ you meant: dump enabled _but_ failed to actually enable it ]

Comment (rgs1) [reply test/server/server_test.cc:1337]: I did have to make `PlatformImpl` a unique_ptr passable via the constructor, to be able to make `enableCoreDump()` fail... hopefully it's not too bad.

Comment (rgs1) [reply test/server/server_test.cc:1337]: Err, I thought you mean a test case for dump enabled but prctl call fails -- so I added a test for that. Which I think is good, we want that path too.

Note that, there is no warning when core dump is not enabled... only when it's enabled but we fail to actually enabled. Did you initially meant you wanted a warn when it's not enabled? [ see the changes i just pushed ]

Comment (mattklein123) [reply test/server/server_test.cc:1337]: Sorry yes I mistyped.

File: bazel/envoy_library.bzl
Comment (lizan) [inline bazel/envoy_library.bzl:231]: We should refactor this a bit with `config_settings_group`: https://github.com/bazelbuild/bazel-skylib/blob/master/docs/selects_doc.md

File: source/exe/main_common.h
Comment (mattklein123) [inline source/exe/main_common.h:41]: unique_ptr?

File: test/exe/main_common_test.cc
Comment (mattklein123) [inline test/exe/main_common_test.cc:141]: Can you actually expect the call and then also use EXPECT_LOG_CONTAINS to check for the warning?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Few more comments and then LGTM modulo Lizan build comments.

/wait

File: source/exe/main_common.h
Comment (rgs1) [reply source/exe/main_common.h:41]: oops yeah mistake

File: test/exe/main_common_test.cc
Comment (rgs1) [reply test/exe/main_common_test.cc]: done

File: bazel/envoy_library.bzl
Comment (rgs1) [reply bazel/envoy_library.bzl:231]: actually, I noticed that on some places (e.g.: for quiche) we just match on:

```
"@envoy//bazel:linux"
```

whereas in most other cases we explicitly enumerate each Linux arch, when is one way preferred over the other?

Comment (lizan) [reply bazel/envoy_library.bzl:231]: If that works go for it. Sometime we have to differentiate architecture but this is not the case.

Comment (rgs1) [reply bazel/envoy_library.bzl:231]: Yeah I think it does, I had it that way initially and then I switched over when debugging something. Let me try.

File: test/exe/main_common_test.cc
Comment (mattklein123) [inline test/exe/main_common_test.cc:144]: nit: you can just add a MOCK_METHOD here that allows you to have a normal expectation, etc.?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small comment.

/wait

File: test/exe/main_common_test.cc
Comment (rgs1) [reply test/exe/main_common_test.cc]: sure done

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Review by lizan — APPROVED** [review]

**Pull request merged**
